text/0000-injection-hook-normalization.md

```diff
@@ -0,0 +1,302 @@
+
+## How we teach this
+
+We would teach these concepts along with our material on DI in general.
+`didCreate` and its counter-part, `willDestroy`, should be thought of as
+container hooks that can be used to setup and teardown state based on
+injections.
+
+An important point to cover will be the differences between `constructor` and
+`didCreate`. We should teach `constructor` specifically as a mechanism for
+_shaping_ the class, defining the basic fields and properties that will exist on
+it. By contrast, `didCreate` is for setting up state within the larger system -
+the Ember application. Concrete examples can be used to drive this point home.
+
+## Drawbacks
+
+- The previously proposed Glimmer component API has been built and highly
+  publicized. This change will definitely be surprising to some people, and may
+  cause confusion. However, v1.0.0 of Glimmer component has _not_ been published
+  yet, and we would not be breaking any semver guarantees. Additionally, there
+  are likely not many early adopters, since we have not yet marked it as stable.
+
+- There has been a lot of concern about confusion between `constructor` and
+  `init` as we head further down the native class adoption path. This decision
+  would formalize adopting `didCreate`, which is very similar to `init`, for
+  _all_ classes for the forseeable future. This could also cause confusion,
+  since container based classes would essentially always have this dichotomy of
+  two "creation" hooks.
+
+  Creating a different hook with a new name should help with this issue. `init`
+  has historically been taught as the same as `constructor`, which is why some
+  users assume it will have the exact same semantics, and are confused when it
+  does not. Documenting the purpose for `didCreate`, along with clear, practical
+  examples for when to use it, and when to use `constructor`, should go a long
+  way toward preventing any additional confusion.
+
+## Alternatives
+
+- Standardizing on always passing the `owner` as the first parameter is another
+  option here. This would bring us inline with the current behavior of Glimmer
+  components, and a short term solution that would enable this is proposed in
+  the [Classic Class Owner Tunnel RFC](https://github.com/emberjs/rfcs/pull/451).
+
+- This RFC approaches the current problem directly - how do we address the
+  differences in behaviors of injections between different base classes in
+  Ember today, and what is the way we want users to interact with injections in
+  the long run?
+
+  However, it is also touching on some larger underlying infrastructural debt.
+  The reason this has become an issue at all is, in part, because Ember's
+  container is heavily tied to the `EmberObject` model and its implementation
+  details. There are several ways we could move forward with disconnecting
+  these:
+
+  1. We could move the container in a more annotation based direction, as
```

Comment:
I'm sure a lot of thought has gone into the form factor here, but I'm going to offer my less-deeply-considered 2¢ anyway 🙂

My gut reaction to this proposal is that normalizing injection across the board is excellent for API consistency, but I have a fairly strong preference for something akin to `@start` and `@teardown` over `didCreate` and `willDestroy`.

One of the things I really like about the Glimmer component API is that it relieves a lot of the cognitive burden that the classic API placed on developers in the form of special method names they needed to be aware of. Moving those things to be explicit opt-in annotations in the template eliminates a lot of implicit "magic" and makes it much easier for someone new to Ember to understand why a particular method is being called.

Explicitly annotating methods that should be invoked as part of the container lifecycle feels like the moral container equivalent of the move to render modifiers for component lifecycle events. Putting aside the hypothetical `@glimmer` and constructor parameter `@service` decorators (which I agree would be a much bigger shift), it feels like `@start` and `@teardown` could fit neatly into the injection system we have today?